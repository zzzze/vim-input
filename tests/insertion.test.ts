import { describe, expect, it } from 'vitest'
import { ElementController } from './utils/element-controller'
import { prepare } from './utils/prepare'
import delay from './utils/delay'

describe('insertion (o)', async () => {
  it('should append new line after current line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('o')
    expect(element.value).toMatchSnapshot()
    await delay(200)
    expect(dom.window.app.vim.currentMode).to.equal('edit_mode')
  })

  it('should append new line before current line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('shift+o')
    expect(element.value).toMatchSnapshot()
    await delay(200)
    expect(dom.window.app.vim.currentMode).to.equal('edit_mode')
  })
})

describe('insertion (a)', async () => {
  it('should append characters in most front of empty line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('j')
    await controller.pressKey('a')
    await delay(200)
    await controller.input('hello world')
    expect(element.value).toMatchSnapshot()
  })

  it('should restore empty line placeholder if nothing appended to the empty line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('j')
    const selectionStart = element.selectionStart
    const selectionEnd = element.selectionEnd
    await controller.pressKey('a')
    await delay(200)
    await controller.pressKey('Escape')
    await delay(200)
    expect(element.value).toMatchSnapshot()
    expect(element.selectionStart).to.equal(selectionStart)
    expect(element.selectionEnd).to.equal(selectionEnd)
  })
})
