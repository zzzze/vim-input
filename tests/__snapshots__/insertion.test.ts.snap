// Vitest Snapshot v1

exports[`insertion (a) > should append characters in most front of empty line 1`] = `
"Dear John,
hello world
I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;

exports[`insertion (a) > should restore empty line placeholder if nothing appended to the empty line 1`] = `
"Dear John,
 
I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;

exports[`insertion (o) > should append new line after current line 1`] = `
"Dear John,
 

I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;

exports[`insertion (o) > should append new line before current line 1`] = `
" 
Dear John,

I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;
